include/seqan3/core/debug_stream/default_printer.hpp

```diff
@@ -0,0 +1,224 @@
+     */
+    template <typename stream_t,
+              typename arg_t,
+              std::ptrdiff_t i = find_index<stream_t, arg_t, printer_templates_t<arg_t>...>()>
+    using printer_for_t =
+        // first: the index of the printer that can print the arguments or sizeof...(printers_t) if no printer was found
+        // second: the tuple of instantiated printers extended with no_printer_found
+        std::tuple_element_t<i, std::tuple<printer_templates_t<arg_t>..., no_printer_found>>;
+};
+
+/*!
+ * \brief The default printer that is used by seqan3::debug_stream.
+ *
+ * This printer offers a function call operator if the type is printable by one of the printers defined
+ * in the seqan3::printer_order.
+ * If no valid printer is found the function call operator is not defined.
+ *
+ * \ingroup core_debug_stream
+ */
+struct default_printer :
+    public printer_order<
+        debug_stream_printer,
+        alignment_result_printer,                 // type seqan3::alignment_result<>
+        alignment_printer,                        // concept seqan3::tuple_like<>
+        advanceable_alignment_coordinate_printer, // type seqan3::detail::advanceable_alignment_coordinate<>
+        alignment_matrix_printer,                 // concept seqan3::detail::matrix<>
+        trace_directions_printer,                 // type seqan3::detail::trace_directions
+        mask_printer,                             // type seqan3::mask
+        integral_printer,                         // concept std::integral
+        // NOTE: alphabet_printer needs the integral_printer overload, otherwise it might have infinite recursion due to
+        // char and uint being an alphabet
+        alphabet_printer,         // concept seqan3::alphabet
+        char_sequence_printer,    // concept std::range::input_range<> with char value_type
+        integer_sequence_printer, // concept std::range::input_range<> with std::integral value_type
+        sequence_printer, // concept seqan3::sequence<>, i.e. std::range::input_range<> with seqan3::alphabet value_type
+        input_range_printer, // concept std::range::input_range<>
+        optional_printer,    // type std::optional<> or std::nullopt_t
+        tuple_printer,       // concept seqan3::tuple_like<>
+        std_printer          // anything that can be printed by std::ostream
+        >
+{
+public:
+    /*!
+     * \brief The function call operator that is only defined if the type is printable.
+     * \tparam stream_t The type of the stream.
+     * \tparam arg_t The type of the argument.
+     * \param[in,out] stream The stream to print to.
+     * \param[in] arg The argument to print.
+     */
+    template <typename stream_t, typename arg_t>
+        requires printable_with<printer_for_t<stream_t &, arg_t>, stream_t &, arg_t>
+    constexpr void operator()(stream_t & stream, arg_t && arg) const
+    {
+        using printer_t = printer_for_t<stream_t &, arg_t>;
```

Comment:
```suggestion
        using printer_t = printer_for_t<stream_t &, std::remove_cvref_t<arg_t>>;
```

If we did this here and in the `printable_with` concept, we might have to do it in every printer class?

Reply:
I don't think removing cvref is right here, since we simply test whether the printer type is invocable with the stream type and the forwarded argument. This class merely forwards the argument to the first viable printer, as long as it is invocable with that forwarded argument. 

Regarding the `printable_with` question; this class is specific to seqan3 and just needs to check whether a given printer can be invoked in order to find the first viable printer. The actual printers don't need to check whether they are printable, since they know what they can print and what not. However, inside of the debug stream streaming operator we still need to decide whether we use the older interface with the debug_stream overloads or the new printer design. Thus, we need to check again whether the printer composite is also printable. I hope thid makes sense?